# Notebook Contents
01. Importing Libraries
02. Importing Data
03. Creating User-defined Functions
04. If-Statements with the loc() Function
05. If-Statements with For-Loops
06. Task 4.7

# 01. Importing Libraries

In [5]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [6]:
#Creating path
path =r'C:\Users\smflo\Documents\03-2023 Instacart Basket Analysis'

In [7]:
df_ords_prod_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# 03. Creating User-defined Functions

In [8]:
#Creating subset first 1000000 rows
df = df_ords_prod_merged[:1000000]

In [9]:
#Checking df shape
df.shape

(1000000, 14)

In [10]:
#Defining the function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [11]:
#Applying the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\smflo\AppData\Local\Temp\ipykernel_22784\3959542531.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
#Checking value counts
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [13]:
#Checking most expensive product in df
df['prices'].max()

14.8

# 04. If-Statements with the loc() Function

In [40]:
#Creating condition
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [15]:
#Creating condition
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [16]:
#Creating condition
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
#Checking value counts
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [18]:
#Creating condition
df_ords_prod_merged.loc[df_ords_prod_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
#Creating condition
df_ords_prod_merged.loc[(df_ords_prod_merged['prices'] <= 15) & (df_ords_prod_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [20]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
#Checking value counts
df_ords_prod_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

# 05. If-Statements with For-Loops

In [22]:
#Checking value counts
df_ords_prod_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [33]:
#Creating for-loop
result = []

for value in df_ords_prod_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [34]:
#Viewing output of for-loop
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [35]:
#Combining result with df_ords_prod_merged
df_ords_prod_merged['busiest_day'] = result

In [36]:
#Checking value counts
df_ords_prod_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# 06. Task 4.7

## Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [31]:
#Creating for-loop
result_2 = []

for value in df_ords_prod_merged["orders_day_of_week"]:
  if value == 0:
    result_2.append("Busiest days")
  elif value == 1:
    result_2.append("Busiest days")
  elif value == 4:
    result_2.append("Slowest days")
  elif value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [32]:
#Combining result with df_ords_prod_merged
df_ords_prod_merged['busiest_day'] = result_2

In [44]:
#Viewing first 5 rows with added column
df_ords_prod_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

## Check the values of this new column for accuracy. Note any observations in markdown format.

In [29]:
#Checking value counts
df_ords_prod_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_day, dtype: int64

In [30]:
#Checking value counts
df_ords_prod_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

#If you add days 0 and 1 together, you get 11864412 same as value count for busiest days. If you add days 3 and 4 together, you get 7624336 same as value count for slowest days.

## When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [38]:
#Checking value counts
df_ords_prod_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [42]:
#Creating list showing percentiles for value counts of column 'order_hour_of_day'
orders = list(df_ords_prod_merged['order_hour_of_day'].value_counts().quantile([0.25, 0.5, 0.75]))

In [43]:
#Viewing percentiles
orders

[272562.0, 1117230.5, 2556034.5]

In [60]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 0, 'busiest_period_of_day'] = 'Fewest orders'

In [61]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 1, 'busiest_period_of_day'] = 'Fewest orders'

In [62]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 5, 'busiest_period_of_day'] = 'Fewest orders'

In [63]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 2, 'busiest_period_of_day'] = 'Fewest orders'

In [64]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 4, 'busiest_period_of_day'] = 'Fewest orders'

In [65]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 3, 'busiest_period_of_day'] = 'Fewest orders'

In [66]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 12, 'busiest_period_of_day'] = 'Most orders'

In [67]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 13, 'busiest_period_of_day'] = 'Most orders'

In [68]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 14, 'busiest_period_of_day'] = 'Most orders'

In [69]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 15, 'busiest_period_of_day'] = 'Most orders'

In [70]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 10, 'busiest_period_of_day'] = 'Most orders'

In [71]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 11, 'busiest_period_of_day'] = 'Most orders'

In [72]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 6, 'busiest_period_of_day'] = 'Average orders'

In [73]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 7, 'busiest_period_of_day'] = 'Average orders'

In [74]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 8, 'busiest_period_of_day'] = 'Average orders'

In [75]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 9, 'busiest_period_of_day'] = 'Average orders'

In [76]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 16, 'busiest_period_of_day'] = 'Average orders'

In [77]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 17, 'busiest_period_of_day'] = 'Average orders'

In [78]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 18, 'busiest_period_of_day'] = 'Average orders'

In [79]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 19, 'busiest_period_of_day'] = 'Average orders'

In [80]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 20, 'busiest_period_of_day'] = 'Average orders'

In [81]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 21, 'busiest_period_of_day'] = 'Average orders'

In [82]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 22, 'busiest_period_of_day'] = 'Average orders'

In [83]:
#Creating Condition
df_ords_prod_merged.loc[df_ords_prod_merged['order_hour_of_day'] == 23, 'busiest_period_of_day'] = 'Average orders'

## Print the frequency for this new column.

In [84]:
#Checking value counts
df_ords_prod_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       16128666
Average orders    15679865
Fewest orders       596328
Name: busiest_period_of_day, dtype: int64

In [85]:
#view first 5 rows
df_ords_prod_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day busiest_period_of_day  
0  Regularly busy        Average orders  
1  Regularly busy        Average orders  
2  Regularly busy           Most orders  
3      Least busy        Average orders  
4      Least busy           Most orders

# 07. Exporting Data

In [86]:
#Export to .pkl
df_ords_prod_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))